### PR TITLE
FIX: don't error topic RSS when posts are deleted

### PR DESCRIPTION
**Pull request by lis2**
Crawlers should not trigger an error and return 404 when posts are soft deleted

### CODE REVIEW

File: spec/requests/topics_controller_spec.rb [hunk 1]
```diff
@@ -2512,6 +2512,13 @@ def extract_post_stream
       expect(response.media_type).to eq('application/rss+xml')
     end
 
+    it 'renders rss even if post is deleted' do
+      topic.posts.map(&:destroy)
```
Comment (CvX) [inline spec/requests/topics_controller_spec.rb:2516]: The PR description says "soft deleted" (meaning `.trash!`?) but the spec here is performing `.destroy`. Should one of them be changed?

**Comment by ZogStriP** [comment]
Shouldn't we reply a 404 if the post has been deleted instead of rendering an incomplete RSS "feed"?

**Comment by lis2** [comment]
@ZogStriP good point, I changed it to return 404 :)
@CvX you are right, change to 🗑️

File: app/controllers/topics_controller.rb [hunk 2]
```diff
@@ -842,6 +842,7 @@ def timings
   def feed
     @topic_view = TopicView.new(params[:topic_id])
     discourse_expires_in 1.minute
+    raise Discourse::NotFound if @topic_view.posts.blank?
```
Comment (ZogStriP) [inline app/controllers/topics_controller.rb:845]: Loading the whole `TopicView` might be "heavy" one the db. I wonder if there's an easy way to make a tiny check that there's at least one post before loading the `TopicView`?

File: app/controllers/topics_controller.rb [hunk 3]
```diff
@@ -840,9 +840,10 @@ def timings
   end
 
   def feed
+    raise Discourse::NotFound if Post.where(topic_id: params[:topic_id]).count == 0
```
Comment (ZogStriP) [inline app/controllers/topics_controller.rb:843]: You can do much shorter using `exists?` 😉 

```suggestion
    raise Discourse::NotFound unless Post.exists?(topic_id: params[:topic_id])
```

**Review by ZogStriP — APPROVED** [review]

File: spec/requests/topics_controller_spec.rb [hunk 4]
```diff
@@ -2519,6 +2519,12 @@ def extract_post_stream
       expect(response.body).to_not include("/forum/forum")
       expect(response.body).to include("http://test.localhost/forum/t/#{topic.slug}")
     end
+
+    it 'returns 404 when posts are deleted' do
+      topic.posts.map(&:trash!)
```
Comment (ZogStriP) [inline spec/requests/topics_controller_spec.rb:2524]: You don't need to `map` when you don't use the return value 😉 

```suggestion
      topic.posts.each(&:trash!)
```

**Pull request merged**